### ISSUE TITLE
Create an Android Studio Plugin for Testify

### ISSUE THREAD

**Issue by DanielJette** [Shopify/android-testify#161]
# Android Studio Plugin for Testify

## TL:DR

Create a native plugin for Testify to fully integrate all the key Testify commands directly into Android Studio.

## What?

Develop an Intellij Platform Plugin suitable for use in Android Studio to seamlessly integrate the full Testify feature set directly into the IDE.

## Why?

Testify screenshot tests are built on top of Android Instrumentation tests and so already integrate seamlessly with existing test suites. Screenshots can be captured directly from within Android Studio or using the Gradle command-line tools.

However, the current Android Studio support relies fully on the fact that Testify tests extend `ActivityTestRule` and can be invoked using the built-in support for running instrumentation tests with various commands (notably sidebar icons) in Android Studio. These are limited to `run` and `debug` commands. Android Studio has no concept of recording or pulling screenshots from the device. Thus, it requires developers to drop to the Gradle Panel or command-line to fully use Testify. This project will enhance the developer experience by adding fully integrated IDE UI for all relevant Testify commands.

## How?

Products based on the IntelliJ Platform can be modified and adjusted for custom purposes by adding plugins. Tool integration makes it possible to manipulate third-party tools and components directly from the IDE without switching contexts.

This implies:

- Implementation of additional actions
- Related UI components
- Access to external resources

Android Studio plugins extend or add functionality to the Android Studio IDE. Plugins can be written in Kotlin or Java, or a mix of both, and are created using IntelliJ IDEA and the IntelliJ Platform.

## Resources

- [Testify Source](http://shopify.github.io/android-testify)
- [ActivityTestRule](https://developer.android.com/reference/androidx/test/rule/ActivityTestRule)
- [Android Studio Plugin Development](https://www.jetbrains.org/intellij/sdk/docs/products/android_studio.html)
- [Creating Your First Plugin](https://www.jetbrains.org/intellij/sdk/docs/basics/getting_started.html)
- [IntelliJ Platform Plugin Template](https://github.com/JetBrains/intellij-platform-plugin-template)

**Comment by DanielJette** [comment]
## To-do list

- [ ] Fork the Intellij Platform Plugin template
- [ ] Update the template to reference Testify
- [ ] Create a custom icon
- [ ] Identify the `@ScreenshotInstrumentation` annotation on test methods
- [ ] Draw a gutter icon on the line next to `@ScreenshotInstrumentation`
- [ ] Handle clicks on the icon
- [ ] Display a popup-menu displaying the name of the method being clicked on
- [ ] Access `adb` commands or custom gradle tasks from the existing plugin
- [ ] Menu for test
- [ ] Menu for record
- [ ] Menu for clear
- [ ] Menu for delete
- [ ] Menu for pull

**Comment by DanielJette** [comment]
## Alpha Prep

- [ ] Add license
- [ ] Update readme
- [ ] Add contribution doc
- [ ] Remove java version
- [ ] Remove unused actions/listeners
- [ ] Refactor package hierarchy
- [ ] Optimize qualifications & parsing

## Release

- [ ] Review template instructions

**Comment by DanielJette** [comment]
Provisional listing: https://plugins.jetbrains.com/plugin/14687-android-testify--screenshot-instrumentation-tests/

[Testify.IntelliJ.Plugin-0.0.1-alpha.zip](https://github.com/Shopify/android-testify/files/4903481/Testify.IntelliJ.Plugin-0.0.1-alpha.zip)

**Comment by DanielJette** [comment]
Included in 1.0.0
